rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -853,42 +855,79 @@ trait PromiseGeneration_updateAndWriteRule extends PromiseGenerationService {
   }
 
   /**
+   * For each nodeConfiguration, get the corresponding node config version.
+   * Either get it from cache or create a new one depending if the node configuration was updated
+   * or not.
+   */
+  def getNodesConfigVersion(allNodeConfigs: Map[NodeId, NodeConfiguration], hashes: Map[NodeId, NodeConfigurationHash], generationTime: DateTime): Map[NodeId, NodeConfigId] = {
+
+    /*
+     * Several steps heres:
+     * - look what node configuration are updated (based on their cache ?)
+     * - write these node configuration
+     * - update caches
+     */
+    val updatedNodes = nodeConfigurationService.selectUpdatedNodeConfiguration(allNodeConfigs, hashes)
+
+    /*
+     * The hash is directly the NodeConfigHash.hashCode, because we want it to be
+     * unique to a given generation and the "writtenDate" is part of NodeConfigurationHash.
+     * IE, even if we have the same nodeConfig than a
+     * previous one, but the expected node config was closed, we want to get a new
+     * node config id.
+     */
+    def hash(h: NodeConfigurationHash): String = {
+      //we always set date = 0, so we have the possibility to see with
+      //our eyes (and perhaps some SQL) two identicals node config diverging
+      //only by the date of generation
```

Comment:
If only the date is different between two config, 

date1-nodeConfigHash and date2-nodeConfigHash

if you do not set the same date when using hashcode, the hash will be different and you will have 

date1-nodeConfigHash1 and date2-nodeConfigHash2

so you can't know that they have only the date that is different